File: CURE/CURE.py
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
Philippe M. Noël
Clustering Using REpresentatives (CURE) Algorithm -- Python 3
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
import numpy as np
import scipy as scipy
import matplotlib.pyplot as plt
from numpy import linalg
from scipy import io
from scipy.spatial.distance import pdist,squareform
from scipy.cluster.hierarchy import dendrogram, linkage
from scipy.cluster.hierarchy import fcluster

def CURE():
    """ Clustering Using REpresentatives from .mat file """
    # loading data
    X = scipy.io.loadmat('CURE_data.mat')['X']
    N = len(X[1,:])

    # randomly submsapling data to generate representatives
    np.random.seed(3989) # other seeds could be used
    X = X[:,np.random.permutation(N)]
    N_reps = 300 # arbitrarily subsampling first 300 points
    X_reps = X[:,0:N_reps]

    # hierarchical clustering of subsampe to get representatives
    Z = linkage(X.T, 'single') # single linkage due to nature of data
    k = 2 # 2 clusters
    IDX_reps = fcluster(Z, k, criterion='maxclust')

    # plotting representatives for visualization
    plt.figure(1); plot.clf()
    reds = np.vstack([X_reps[:,i] for i in range(N_reps) if IDX_reps[i] == 1])
    blues = np.vstack([X_reps[:,i] for i in range(N_reps) if IDX_reps[i] == 2])
    plt.plot(reds[:,0], red[:,1], '.r'); plt.plot(blues[:,0], blues[:,1], '.b')
    plt.legend(['Cluster #1', 'Cluster #2'])
    plt.title('Representatives, N = ' + str(N_reps))

    # going over data set to assign all points to clusters
    distances, IDX = np.zeros(N_reps), np.zeros(N)
    for i in range(N):
        # find nearest representative to data point i
        for j in range(N_reps):
            distances[j] = scipy.linalg.norm(X[:,i] - X_reps[:,j])
        # find nearest cluster and assign it
        nearest = np.argmin(distances)
        IDX[i] = IDX_reps[nearest]

    # plotting clustered dataset
    plt.figure(2); plot.clf()
    reds = np.vstack([X[:,i] for i in range(N) if IDX[i] == 1])
    blues = np.vstack([X[:,i] for i in range(N) if IDX[i] == 2])
    plt.plot(reds[:,0], red[:,1], '.r'); plt.plot(blues[:,0], blues[:,1], '.b')
    plt.legend(['Cluster #1', 'Cluster #2'])
    plt.title('Full Dataset, N = ' + str(N))


# test driver
CURE()

File: CURE/CURE.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Philippe M. Noël
% Clustering Using REpresentatives (CURE) Algorithm -- Matlab
% Original Code from Harvard APMTH120
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% loading & preparing data
X = X; % must load data from .mat file by double-clicking on it 
N = length(X(1,:));
rng(3989); % randomly shuffling data
X = X(:,randperm(N));

%% clustering N representatives using Hierarchical Clustering
N_reps = 300;
X_reps = X(:,1:N_reps);
Y = pdist(X_reps','euclid'); % distances
Z = linkage(Y,'single');
k = 2; % 2 clusters
IDX_reps = cluster(Z,'maxclust',k)

%% cluster all points by going over full dataset
for i = 1:N
  % find nearest representative to data point i
  for j = 1:N_reps
    distances(j) = norm(X(:,i) - X_reps(:,j));
  end
  % assign data point i to cluster of nearest representative point
  [M,nearest] = min(distances);
  IDX(i) = IDX_reps(nearest);
end

%% plot representatives
set(0,'defaulttextfontsize',18); set(0,'defaultaxesfontsize',18);
figure(1); clf
red_rep = X_reps(:,IDX_reps == 1); plot(red_rep(1,:),red_rep(2,:),'.r');
hold on
blue_rep = X_reps(:,IDX_reps == 2); plot(blue_rep(1,:),blue_rep(2,:),'.b');
title(sprintf('Representatives, N = %d', N_reps));
legend('Cluster #1','Cluster #2');


%% plot clustered data
figure(2); clf
red = X(:,IDX == 1); blue = X(:,IDX == 2);
plot(red(1,:),red(2,:),'.r');
hold on
plot(blue(1,:),blue(2,:),'.b');
title(sprintf('Full Data, N = %d',N));
legend('Cluster #1','Cluster #2');
